### DISCUSSION TITLE
How to protect routes using @supabase/ssr? (NextJS)

### DISCUSSION THREAD

**Discussion by OrhanTozan** [supabase/supabase#21468] [Questions]
I think I need to put something into middleware.ts. But the starter template already has code in middleware.ts. Im afraid that my code will hurt the existing code.

**Comment by ryanhalliday** [comment]
Put this in the route I suppose

```ts
  const {
    data: { user },
  } = await supabase.auth.getUser();

  if (!user) {
    return redirect("/login");
  }
```

You can see the full example here: https://github.com/vercel/next.js/blob/canary/examples/with-supabase/app/protected/page.tsx#L11-L17

---

If you want to do it for certain routes from in the middleware you can put that after `await updateSession(request)` from the Supabase example. Next.js example here: https://nextjs.org/docs/app/building-your-application/routing/middleware#conditional-statements

**Reply by OrhanTozan** [reply]
But updateSession(request) returns a response. What should I do with it?

**Reply by ryanhalliday** [reply]
Next.js middlewares are honestly a mess, terrible design... 

You would need to effectively merge all your middlewares together into one big middleware. It may be easier if you have them run on certain conditions and just check certain routes first, eg https://nextjs.org/docs/messages/nested-middleware

The Supabase middleware could quite easily be modified to take in a request and response, modify those as it needs, and return them back.

**Reply by guillermoscript** [reply]
Hey @OrhanTozan I did this and its working just fine, hope this helps:
```ts
// @/supabase/middleware.ts
import { createServerClient, type CookieOptions } from '@supabase/ssr'
import { NextResponse, type NextRequest } from 'next/server'

export async function updateSession(request: NextRequest) {
    let response = NextResponse.next({
        request: {
            headers: request.headers,
        },
    })

    const supabase = createServerClient(
        process.env.NEXT_PUBLIC_SUPABASE_URL!,
        process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,
        {
            cookies: {
                get(name: string) {
                    return request.cookies.get(name)?.value
                },
                set(name: string, value: string, options: CookieOptions) {
                    request.cookies.set({
                        name,
                        value,
                        ...options,
                    })
                    response = NextResponse.next({
                        request: {
                            headers: request.headers,
                        },
                    })
                    response.cookies.set({
                        name,
                        value,
                        ...options,
                    })
                },
                remove(name: string, options: CookieOptions) {
                    request.cookies.set({
                        name,
                        value: '',
                        ...options,
                    })
                    response = NextResponse.next({
                        request: {
                            headers: request.headers,
                        },
                    })
                    response.cookies.set({
                        name,
                        value: '',
                        ...options,
                    })
                },
            },
        }
    )

    const user = await supabase.auth.getUser()

    if (request.nextUrl.pathname.startsWith('/dashboard') && user.error) {
        return NextResponse.redirect(new URL('/auth/login', request.url))
    }


    return response
}
```

and my middleware:
```ts
// middleware.ts
import { type NextRequest } from 'next/server'
import { updateSession } from '@/lib/supabase/middleware'

export async function middleware(request: NextRequest) {
    return await updateSession(request)
}

export const config = {
    matcher: [
        /*
         * Match all request paths except for the ones starting with:
         * - _next/static (static files)
         * - _next/image (image optimization files)
         * - favicon.ico (favicon file)
         * Feel free to modify this pattern to include more paths.
         */
        '/((?!_next/static|_next/image|favicon.ico|.*\\.(?:svg|png|jpg|jpeg|gif|webp)$).*)',
    ],
}
```
The dashboard page its only for authenticated users and this works great for me

**Reply by Justinfan827** [reply]
If you've called getUser in the middleware, do we still need to be validating users in server components? Or can we just call getSession to skip the network call?

Seems like we'd still need to call getUser again in a nextjs page to authenticate the user?

https://supabase.com/docs/guides/auth/server-side/nextjs?queryGroups=router&router=app

<img width="1522" alt="image" src="https://github.com/supabase/supabase/assets/22778148/aa63fdb7-c126-4dac-a589-4fd8701f93be">

**Reply by OrhanTozan** [reply]
I am using getSession() inside the middleware now. The reason for this, assuming your data is secured by RLS, you dont need extra security, and getUser() just slows things down.

Yes, protecting Page A with a supabase session would allow users that spoof their cookies to access that page, but if that page makes calls to tables with RLS turned on, the user would see no data. So there is nothing to worry about.

**Reply by matsyui** [reply]
didnt work for me sad.

**Reply by e-afzal** [reply]
> Hey @OrhanTozan I did this and its working just fine, hope this helps:
> ```ts
> // @/supabase/middleware.ts
> import { createServerClient, type CookieOptions } from '@supabase/ssr'
> import { NextResponse, type NextRequest } from 'next/server'
> 
> export async function updateSession(request: NextRequest) {
>     let response = NextResponse.next({
>         request: {
>             headers: request.headers,
>         },
>     })
> 
>     const supabase = createServerClient(
>         process.env.NEXT_PUBLIC_SUPABASE_URL!,
>         process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,
>         {
>             cookies: {
>                 get(name: string) {
>                     return request.cookies.get(name)?.value
>                 },
>                 set(name: string, value: string, options: CookieOptions) {
>                     request.cookies.set({
>                         name,
>                         value,
>                         ...options,
>                     })
>                     response = NextResponse.next({
>                         request: {
>                             headers: request.headers,
>                         },
>                     })
>                     response.cookies.set({
>                         name,
>                         value,
>                         ...options,
>                     })
>                 },
>                 remove(name: string, options: CookieOptions) {
>                     request.cookies.set({
>                         name,
>                         value: '',
>                         ...options,
>                     })
>                     response = NextResponse.next({
>                         request: {
>                             headers: request.headers,
>                         },
>                     })
>                     response.cookies.set({
>                         name,
>                         value: '',
>                         ...options,
>                     })
>                 },
>             },
>         }
>     )
> 
>     const user = await supabase.auth.getUser()
> 
>     if (request.nextUrl.pathname.startsWith('/dashboard') && user.error) {
>         return NextResponse.redirect(new URL('/auth/login', request.url))
>     }
> 
> 
>     return response
> }
> ```
> 
> and my middleware:
> ```ts
> // middleware.ts
> import { type NextRequest } from 'next/server'
> import { updateSession } from '@/lib/supabase/middleware'
> 
> export async function middleware(request: NextRequest) {
>     return await updateSession(request)
> }
> 
> export const config = {
>     matcher: [
>         /*
>          * Match all request paths except for the ones starting with:
>          * - _next/static (static files)
>          * - _next/image (image optimization files)
>          * - favicon.ico (favicon file)
>          * Feel free to modify this pattern to include more paths.
>          */
>         '/((?!_next/static|_next/image|favicon.ico|.*\\.(?:svg|png|jpg|jpeg|gif|webp)$).*)',
>     ],
> }
> ```
> The dashboard page its only for authenticated users and this works great for me
> 

Worked for me! Thanks a lot for this! 😆

**Reply by NicoFerni** [reply]
> 
Hi! I have tried this and its not working, I think that my middleware is not setup properly:

here is my middleware in my root file

``` import { type NextRequest } from 'next/server'
import { updateSession } from './utils/supabase/middleware'

export async function middleware(request: NextRequest) {
  return await updateSession(request)

}

export const config = {
    matcher: [
        /*
         * Match all request paths except for the ones starting with:
         * - _next/static (static files)
         * - _next/image (image optimization files)
         * - favicon.ico (favicon file)
         * Feel free to modify this pattern to include more paths.
         */
        '/((?!_next/static|_next/image|favicon.ico|.*\\.(?:svg|png|jpg|jpeg|gif|webp)$).*)',
    ],
}

```
and my supabase middleware:

```
import { createServerClient, type CookieOptions } from '@supabase/ssr'
import { NextResponse, type NextRequest } from 'next/server'

export async function updateSession(request: NextRequest) {
    let response = NextResponse.next({
        request: {
            headers: request.headers,
        },
    })
    const supabase = createServerClient(
        process.env.NEXT_PUBLIC_SUPABASE_URL!,
        process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,
        {
            cookies: {
                get(name: string) {
                    return request.cookies.get(name)?.value
                },
                set(name: string, value: string, options: CookieOptions) {
                    request.cookies.set({
                        name,
                        value,
                        ...options,
                    })
                    response = NextResponse.next({
                        request: {
                            headers: request.headers,
                        },
                    })
                    response.cookies.set({
                        name,
                        value,
                        ...options,
                    })
                },
                remove(name: string, options: CookieOptions) {
                    request.cookies.set({
                        name,
                        value: '',
                        ...options,
                    })
                    response = NextResponse.next({
                        request: {
                            headers: request.headers,
                        },
                    })
                    response.cookies.set({
                        name,
                        value: '',
                        ...options,
                    })
                },
            },
        }
    )

    const user = await supabase.auth.getUser()
    if (request.nextUrl.pathname.startsWith('/script') && user.error) {
        return NextResponse.redirect(new URL('/auth/login', request.url))
    }


    return response
}
```

Also I have here my createClient Function

```
import { createServerClient, type CookieOptions } from '@supabase/ssr'
import { cookies } from 'next/headers'


export async function createClient() {
  'use server'
  const cookieStore = await cookies()

  return createServerClient(
    process.env.NEXT_PUBLIC_SUPABASE_URL!,
    process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,
    {
      cookies: {
        get(name: string) {
          return cookieStore.get(name)?.value
        },
        set(name: string, value: string, options: CookieOptions) {
          try {
            cookieStore.set({ name, value, ...options })
          } catch (error) {

          }
        },
        remove(name: string, options: CookieOptions) {
          try {
            cookieStore.set({ name, value: '', ...options })
          } catch (error) {

          }
        },

      },
    }
  )
}
```

I have put some console logs and are not appearing so I think is never running but in my console appear that the middleware compiles

**Reply by alessiotortora** [reply]
are you using src directory in next 14 or higher ? if yes, middleware needs to be in the src where the app directory lives.

**Reply by bilecky** [reply]
working perfectly for me too!
